maven-core/src/main/java/org/apache/maven/lifecycle/internal/builder/BuilderCommon.java

```diff
@@ -105,6 +106,20 @@ public MavenExecutionPlan resolveBuildPlan( MavenSession session, MavenProject p
 
         lifecycleDebugLogger.debugProjectPlan( project, executionPlan );
 
+        // With Maven 4's build/consumer the POM will always rewrite during distribution.
+        // The maven-gpg-plugin uses the original POM, causing an invalid signature.
+        // Fail as long as there's no solution available yet
+        Optional<MojoExecution> gpgMojo = executionPlan.getMojoExecutions().stream()
+                .filter( m -> "org.apache.maven.plugins".equals( m.getGroupId() ) )
+                .filter( m -> "maven-gpg-plugin".equals( m.getArtifactId() ) )
+                .findAny();
+
+        if ( gpgMojo.isPresent() )
+        {
```

Comment:
What about adding a predicate:
```java
private Predicate<MojoExecution> isMavenGpgPlugin = m -> m.getArtifactId().equals(
            "maven-gpg-plugin" ) && m.getGroupId().equals( "org.apache.maven.plugins" );
```
Then you can change the whole filter via the following:
```
        boolean gpgMojo = executionPlan.getMojoExecutions().stream().anyMatch(isMavenGpgPlugin);

        if ( gpgMojo )
        {
            throw new LifecycleExecutionException( "The maven-gpg-plugin is not supported by Maven 4."
                + " Verify if there is a compatible signing solution or use Maven 3" );
        }
```
because the optional is only used as boolean.

Reply:
You might also inline the `gpgMojo` into the if-Statement.

Reply:
To make more clear the combining of G+A: you could go the path:
Defining two predicates like:
```java
    private Predicate<MojoExecution> isMavenGpgGroupId = m -> m.getGroupId().equals( "org.apache.maven.plugins" );
    private Predicate<MojoExecution> isMavenGpgArtifactId = m -> m.getArtifactId().equals( "maven-gpg-plugin" );
```
and later use it like this:
```java
boolean gpgMojo = executionPlan.getMojoExecutions().stream().anyMatch(isMavenGpgGroupId.and(isMavenGpgArtifactId ));
```

Reply:
So yes, there are many options to write this. My style would be: 1 simple statement per filter as it would reduce complexity (TT, TF, FT, FF vs T->T, T->F, F) and no externalized predicates if the statement is short and clear.

Reply:
Let's keep it as simple as possible. Either one or two filters.

Reply:
A separate predicate expresses better what you really mean because it gives it a readable name.
I prefer this one:
```java
boolean gpgMojo = executionPlan.getMojoExecutions().stream().anyMatch(MavenGpgPlugin);
```
Can't be clearer from my POV... If you like more details jump to the predicate via IDE.

Complexity? If you use stream API predicates are more less everywhere. Usually if a method is to long (too complex) you refactor out a part of it with a readable name. Here it's the same.

Furthermore combined filters don't give you the chance to use `anyMatch(..)`, `.allMatch(..)` or `noneMatch(..)` which is often a good choice and often not used based on the usage of chaining several filter(..)...

Reply:
To me the stream is not complex enough to justify this propsed extraction.